HPCC-26155 allow some placement elements appending

xwang2713 opened the pull request:
Both nodeSelector and tolerations entries can be appended to
existing settings. Reference helm/hpcc/docs/placemnets.md for
examples and detail information.

There is an example in test section

Signed off by xiaoming.wang@lexisnexis.com



## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->
'Tested on Azure.
There are three node pools
1) master pool
   No hpcc pod will be scheduled to the master pool
2) node pool (name=npone)
   Most HPCC pods will be placed here except esp and roxie pods
3) node pool (name=nptwo)
   This is spot instance node pool which has taint:
   kubernetes.azure.com/scalesetpriority=spot:NoSchedule
   I add another taint for test: priority=spot:NoSchedule
   esp and roxie pods will be placed here.

```code
kubectl get nodes -o wide
NAME                               STATUS   ROLES   AGE    VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION     CONTAINER-RUNTIME
aks-masternp-40758588-vmss000000   Ready    agent   134m   v1.19.9   10.240.0.4    <none>        Ubuntu 18.04.5 LTS   5.4.0-1048-azure   containerd://1.4.4+azure
aks-masternp-40758588-vmss000001   Ready    agent   134m   v1.19.9   10.240.0.5    <none>        Ubuntu 18.04.5 LTS   5.4.0-1048-azure   containerd://1.4.4+azure
aks-npone-40758588-vmss000000      Ready    agent   131m   v1.19.9   10.240.0.6    <none>        Ubuntu 18.04.5 LTS   5.4.0-1048-azure   containerd://1.4.4+azure
aks-npone-40758588-vmss000001      Ready    agent   131m   v1.19.9   10.240.0.7    <none>        Ubuntu 18.04.5 LTS   5.4.0-1048-azure   containerd://1.4.4+azure
aks-nptwo-40758588-vmss000000      Ready    agent   128m   v1.19.9   10.240.0.8    <none>        Ubuntu 18.04.5 LTS   5.4.0-1048-azure   containerd://1.4.4+azure
aks-nptwo-40758588-vmss000001      Ready    agent   128m   v1.19.9   10.240.0.9    <none>        Ubuntu 18.04.5 LTS   5.4.0-1048-azure   containerd://1.4.4+azure

```

Here is a sample setting in values.yaml
placements:
  - pods: ["all"]
    placement:
      nodeSelector:
        hpcc: "true"
  - pods: ["type:roxie", "type:esp"]
    placement:
      nodeSelector:
        name: "nptwo"
      tolerations:
        - key: "priority"
          operator: "Equal"
          value:  "spot"
          effect: "NoSchedule"
  - pods: ["type:roxie", "type:esp"]
    placement:
      tolerations:
        - key: "kubernetes.azure.com/scalesetpriority"
          operator: "Equal"
          value:  "spot"
          effect: "NoSchedule"
The output for mydali pod:
    spec:
      nodeSelector:
        hpcc: "true"
The output for one of esp pod:
    spec:
      nodeSelector:
        hpcc: "true"
        name: nptwo
      tolerations:
      - effect: NoSchedule
        key: priority
        operator: Equal
        value: spot
      - effect: NoSchedule
        key: kubernetes.azure.com/scalesetpriority
        operator: Equal
        value: spot
     
```code
kubectl get pods -o wide
NAME                                         READY   STATUS    RESTARTS   AGE   IP            NODE                            NOMINATED NODE   READINESS GATES
dfuserver-d7c77df86-h4ss9                    1/1     Running   0          92m   10.244.3.14   aks-npone-40758588-vmss000001   <none>           <none>
eclqueries-6bdd7cd6f5-gx7z8                  1/1     Running   0          92m   10.244.5.4    aks-nptwo-40758588-vmss000000   <none>           <none>
eclscheduler-7547ff6d55-4kgrv                1/1     Running   0          92m   10.244.3.11   aks-npone-40758588-vmss000001   <none>           <none>
eclservices-67dd9499dc-7cnf2                 1/1     Running   0          92m   10.244.5.6    aks-nptwo-40758588-vmss000000   <none>           <none>
eclwatch-54cf9df86d-vh445                    1/1     Running   0          92m   10.244.4.5    aks-nptwo-40758588-vmss000001   <none>           <none>
esdl-sandbox-6ff668759-v8txf                 1/1     Running   0          92m   10.244.5.2    aks-nptwo-40758588-vmss000000   <none>           <none>
hthor-84c7659fd5-bv79m                       1/1     Running   0          92m   10.244.2.8    aks-npone-40758588-vmss000000   <none>           <none>
mydali-59c5bf5f4b-t65sb                      2/2     Running   0          92m   10.244.2.9    aks-npone-40758588-vmss000000   <none>           <none>
myeclccserver-85dbf4c98b-cjv4t               1/1     Running   0          92m   10.244.2.7    aks-npone-40758588-vmss000000   <none>           <none>
roxie-agent-1-7d999856c6-dmrvm               1/1     Running   0          92m   10.244.4.6    aks-nptwo-40758588-vmss000001   <none>           <none>
roxie-agent-1-7d999856c6-g5pw8               1/1     Running   0          92m   10.244.5.5    aks-nptwo-40758588-vmss000000   <none>           <none>
roxie-agent-2-65fbccd5dc-swfsv               1/1     Running   0          92m   10.244.4.3    aks-nptwo-40758588-vmss000001   <none>           <none>
roxie-agent-2-65fbccd5dc-z2dlr               1/1     Running   0          92m   10.244.5.3    aks-nptwo-40758588-vmss000000   <none>           <none>
roxie-toposerver-5fff4ccdb8-464p8            1/1     Running   0          92m   10.244.4.2    aks-nptwo-40758588-vmss000001   <none>           <none>
roxie-workunit-65dff4d78c-nv8np              1/1     Running   0          92m   10.244.3.16   aks-npone-40758588-vmss000001   <none>           <none>
sasha-dfurecovery-archiver-8dcf46f74-f6gm5   1/1     Running   0          92m   10.244.3.15   aks-npone-40758588-vmss000001   <none>           <none>
sasha-dfuwu-archiver-6f68b4647d-nbwct        1/1     Running   0          92m   10.244.2.11   aks-npone-40758588-vmss000000   <none>           <none>
sasha-file-expiry-774d48bdb-2qp5s            1/1     Running   0          92m   10.244.3.12   aks-npone-40758588-vmss000001   <none>           <none>
sasha-wu-archiver-5dbdc5699f-xkprm           1/1     Running   0          92m   10.244.3.13   aks-npone-40758588-vmss000001   <none>           <none>
sql2ecl-fbc5bbbf6-6jkt7                      1/1     Running   0          92m   10.244.4.4    aks-nptwo-40758588-vmss000001   <none>           <none>
thor-eclagent-65b9cdcfbd-9qvrj               1/1     Running   0          92m   10.244.2.10   aks-npone-40758588-vmss000000   <none>           <none>
thor-thoragent-7b4b8b4f7c-7spc4              1/1     Running   0          92m   10.244.3.10   aks-npone-40758588-vmss000001   <none>           <none>
```

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Process of PR-15155, label: HPCC-26155-allow-placements-appending is skipped.
The reason is: It is related to containerised environment only.
Commit ID: 45a26bb870555a7c09096e4f47ac5b438d32984c
OS: centos 7.7.1908 (Linux 3.10.0-1062.9.1.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: hpcc-platform-dev-el7-smoketest-2.novalocal
 Git:  git version 1.8.3.1

xwang2713 commented:
@Greg, please help reviewing docs/placements.md

jakesmith commented on helm/hpcc/docs/placements.md:
```diff
@@ -23,12 +22,55 @@ The list item in "pods" can be one of the following:
 3) affinity
    There is no schema check for the content of affinity. Reference
    https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/
+   Only one "affinity" can be applied to a Pod/Job. If different "affinity" are supplied to multiple pods list which containe a same Pod/Job
```
Should:
```
 If different "affinity" are supplied to multiple pods list which containe a same Pod/Job
```
be:
```
 If a Pod/Job matches multiple placement 'pods' lists, then only the last "affinity" definition will apply.
```

xwang2713 replied:
Fixed

jakesmith commented on helm/hpcc/docs/placements.md:
```diff
@@ -23,12 +22,55 @@ The list item in "pods" can be one of the following:
+   only the last definition prevails.
 4) schedulerName: profile names. "affinity" defined in scheduler profile  requires
    Kubernetes 1.20.0 beta and later releases
+   Only one "scheduerName" can be applied to a Pod/Job.
```
is it true that as with the affinity comment - if there are multiple matches, then only the last will apply?

xwang2713 replied:
Yes

richardkchapman replied:
typo scheduer

jakesmith commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1045,11 +1045,42 @@ Pass in dict with root, category.  optional name to restrict it to a single name
 {{/*
 Create placement related settings
-Pass in dict with placement
+Pass in dict with current placment and new placements dictinary for this entity
```
typo: placment -> placement

xwang2713 replied:
fixed

g-pan replied:
should dictinary be dictionary

xwang2713 replied:
corrected.

jakesmith commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1062,17 +1093,19 @@ Pass in dict with root, job, target and type
 {{- $job := .job -}}
 {{- $target := (printf "target:%s" .target | default "") -}}
 {{- $type := printf "type:%s" .type -}}
+{{- $placementsDict := dict  -}}
```
trivial/formatting: extra space after dict.

xwang2713 replied:
fixed

jakesmith commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1045,11 +1045,42 @@ Pass in dict with root, category.  optional name to restrict it to a single name
 */}}
 {{- define "hpcc.doPlacement" -}}
-{{- if .me.placement }}
-{{ toYaml .me.placement }}
```
to generalize/avoid testing each key, you can express as:
```
{{- if len $me -}}
{{ toYaml .me }}
{{- end -}}
```

xwang2713 replied:
I can't do {{ toYaml .me }} since only need print key, such as "nodeSelector" in the first "nodeSelector" section

Before it was {{ toYaml .me.placement }} but key will be duplicated. For examle,
nodeSelector:
    key1: value1
nodeSelector:
    key2: value2

jakesmith replied:
I may be missing something. Now that addPlacementSetting is called multiple times and merges settings into a single nodeSelector, toleraations etc., and doPlacement is called only once. There will be only 1 nodeSelector key generated afaics.

I tried this change and tested with:
```
placements:
- pods: ["all"]
  placement:
    nodeSelector:
      group: "hpcc"
- pods: ["type:dali"]
  placement:
    nodeSelector:
      spot: "false"
```

and it generates (under dali):
```
    spec:
      nodeSelector:
        group: hpcc
        spot: "false"
```

xwang2713 replied:
Right, that is the limitation. I thought I write it in the placement.md but probably not. I didn't parse the settings before merge since other will need extract each line and get the key. It will be even harder for tolerations since the key can be the same but different handling.  I prefer just add it as limitation since nodeSelector probably will be removed to favor affinity.

xwang2713 replied:
I add this in the placement.md for nodeSelector:
One key can only be applied once for each item in the "pods" list. Otherwise it will result duplicate settings for the same key.

jakesmith replied:
what is wrong with:
```
    spec:
      nodeSelector:
        group: hpcc
        spot: "false"
```

?

jakesmith commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1045,11 +1045,42 @@ Pass in dict with root, category.  optional name to restrict it to a single name
+
+{{/*
+Add placement related settings
+Pass in dict with all placements for this entity
```
should say needs "me":
```
Pass in dict with me, that contains all placements for this entity.
```

xwang2713 replied:
fixed

jakesmith commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1045,11 +1045,42 @@ Pass in dict with root, category.  optional name to restrict it to a single name
-Pass in dict with placement
+Pass in dict with current placment and new placements dictinary for this entity
+*/}}
+{{- define "hpcc.addPlacementSetting" -}}
```
minor: mergePlacementSettings slightly more accurate

jakesmith replied:
Should have a function comment something like 'Pass in dict with me and new'

xwang2713 replied:
fixed

jakesmith commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1045,11 +1045,42 @@ Pass in dict with root, category.  optional name to restrict it to a single name
+{{- $_ := set .new "nodeSelector" (merge (.new.nodeSelector | default dict ) .me.placement.nodeSelector)  }}
+{{- end -}}
+{{- if .me.placement.tolerations }}
+{{- $_ := set .new "tolerations" (concat (.new.tolerations | default list ) .me.placement.tolerations)  }}
```
should this remove duplicates? (i.e. with | uniq)

jakesmith replied:
what happens if there are multiple "pods" matches for tolerations that define the same named key value, meaning that the merged list has multiple entries for the same key(s) ?

xwang2713 replied:
"if or (has $target $placement.pods) (has $type $placement.pods) (has "all" $placement.pods) -}}" will identify which pod/job has the settings.  So it is OK if the same tolerations/nodeSelector is set for multiple pod. But if the same tolerations/nodeSelector  is applied to the same pod more than once the code currently cannot detect it and it will be appended multiple times.

jakesmith replied:
could probably avoid generating duplciates by performing a uniq for tolerations and a deep merge for nodeSelector,
but I think it's ok without doing that, it will mean the generated yaml could have duplicates, but it should still parse/succeed albeit with last duplicate nodeSelector taking effect in that situation.

xwang2713 replied:
I add this as a limitation in placements.md as mentioned in above.

jakesmith reviewed (commented):
@xwang2713 - a few comments

jakesmith commented on helm/hpcc/docs/placements.md:
```diff
@@ -5,11 +5,10 @@ through placement to include an array of objects to configure the Kubernetes Sch
-       placement:
-         <supported configurations>
+   placements:
+   - pods [list]
```
missing colon, should be:
```
- pods: [list]
```

xwang2713 replied:
fixed. There are multiple places

jakesmith commented on helm/hpcc/docs/placements.md:
```diff
@@ -23,12 +22,55 @@ The list item in "pods" can be one of the following:
 1) nodeSelector
+   Multiple nodeSelectors can be applied. For example
+   placements:
+   - pods ["all"]
```
missing colon (and in other places below for 'pods')

xwang2713 replied:
fixed

g-pan commented on helm/hpcc/docs/placements.md:
```diff
@@ -23,12 +22,55 @@ The list item in "pods" can be one of the following:
+     placement:
+       nodeSelector:
+         group: "hpcc"
+   - pods ["type:dali"]
```
think this needs colon after pods

xwang2713 replied:
fixed

g-pan commented on helm/hpcc/docs/placements.md:
```diff
@@ -23,12 +22,55 @@ The list item in "pods" can be one of the following:
+         group: "hpcc"
+         spot: "false"
 2) taints/tolerations
+   Multiple taints/torerations can be applied. For example
```
tolerations

xwang2713 replied:
fixed

g-pan commented on helm/hpcc/docs/placements.md:
```diff
@@ -23,12 +22,55 @@ The list item in "pods" can be one of the following:
 2) taints/tolerations
+   Multiple taints/torerations can be applied. For example
+   placements:
+   - pods ["all"]
```
colon after pods:

xwang2713 replied:
fixed

g-pan commented on helm/hpcc/docs/placements.md:
```diff
@@ -23,12 +22,55 @@ The list item in "pods" can be one of the following:
+       operator: "Equal"
+       value: "hpcc"
+       effect: "NoSchedule"
+   - pods ["type:thor"]
```
colon:

xwang2713 replied:
fixed

g-pan reviewed (commented):
added few comments in line

HPCCSmoketest commented:
Process of PR-15155, label: HPCC-26155-allow-placements-appending starts now.
The reason of this test is: Forced to re-test.
Commit ID: b9110257654dae024ac164c490568669290b8ea2
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-225.ca-central-1.compute.internal
 Git:  git version 2.9.5
Build only for containerized environment.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-225.ca-central-1.compute.internal
 Git:  git version 2.9.5
Sha: b9110257654dae024ac164c490568669290b8ea2
Containerized:True
Conflicting files, should skip build and test.

[Build error:](https://gist.github.com/fa90d865ab58bfa0c7e24107db675b1e)
Build log [Build log](https://gist.github.com/HPCCSmoketest/fa90d865ab58bfa0c7e24107db675b1e#file-result-21-07-21-03-58-38-log)

jakesmith reviewed (commented):
@xwang2713 - looks like this PR need rebasic

Also still not clear on : https://github.com/hpcc-systems/HPCC-Platform/pull/15155/files#r673984750
see my last comment

HPCCSmoketest commented:
Process of PR-15155, label: HPCC-26155-allow-placements-appending starts now.
The reason of this test is: Forced to re-test.
Commit ID: 6bac9a641d533d88c242a5464cd5fd8d14bebfb2
Estimated completion time is 1.09 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-194.ca-central-1.compute.internal
 Git:  git version 2.9.5
Build only for containerized environment.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-194.ca-central-1.compute.internal
 Git:  git version 2.9.5
Sha: 6bac9a641d533d88c242a5464cd5fd8d14bebfb2
Containerized:True
Conflicting files, should skip build and test.

[Build error:](https://gist.github.com/80d7a44737004c5e8645ef691cc786c7)
Build log [Build log](https://gist.github.com/HPCCSmoketest/80d7a44737004c5e8645ef691cc786c7#file-result-21-07-21-19-48-33-log)

xwang2713 commented:
rebased and conflicts are resolved

HPCCSmoketest commented:
Process of PR-15155, label: HPCC-26155-allow-placements-appending starts now.
The reason of this test is: Forced to re-test.
Commit ID: 066176103b99e61002ca3ae33aeb59446618fa4e
Estimated completion time is 1.09 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-112.ca-central-1.compute.internal
 Git:  git version 2.9.5
Build only for containerized environment.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-112.ca-central-1.compute.internal
 Git:  git version 2.9.5
Sha: 066176103b99e61002ca3ae33aeb59446618fa4e
Containerized:True
Build: success
Milestone:Install hpccsystems-platform-community_8.2.5-closedown0.el7.x86_64.rpm
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  734 sec (00:12:14) |  234 sec (00:03:54) |   20 sec (00:00:20) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  997 sec (00:16:37) |

jakesmith reviewed (changes requested):
@xwang2713 - looks like something went wrong with the rebasing, it's currently listed as having 68 commits..

xwang2713 committed:
HPCC-26155 allow some placement elements appending
    Both nodeSelector and tolerations entries can be appended to
    existing settings. Reference helm/hpcc/docs/placemnets.md for
    examples and detail information.

xwang2713 commented:
Replace old branch with a new one against latest candidate-8.2.x

HPCCSmoketest commented:
Process of PR-15155, label: HPCC-26155-allow-placements-appending starts now.
The reason of this test is: Forced to re-test.
Commit ID: bda698a7bd496408a28db7bc31abc0c66d9786ff
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-8.ca-central-1.compute.internal
 Git:  git version 2.9.5
Build only for containerized environment.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-8.ca-central-1.compute.internal
 Git:  git version 2.9.5
Sha: bda698a7bd496408a28db7bc31abc0c66d9786ff
Containerized:True
Build: success
Milestone:Install hpccsystems-platform-community_8.2.7-closedown0.el7.x86_64.rpm
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  505 sec (00:08:25) |  122 sec (00:02:02) |   19 sec (00:00:19) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  655 sec (00:10:55) |

jakesmith reviewed (approved):
@xwang2713 - looks fine/approved, but still wondering re. conclusion to: https://github.com/hpcc-systems/HPCC-Platform/pull/15155/files#r673984750

jakesmith reviewed (commented):
In discussion with @xwang2713 - I think the ordering of the merge dictionaries (e.g. nodeSelector) is wrong or inconsistent with the ordering of other placements.
e.g. the tolerations means only the last will apply, but the nodeSelector merging means the 1st applies.
Changing to mergeOverwrite when it is merging the dictionaries for these items I think will ensure the last will take precendence therefore matching 'last' in other contexts.

Also I think : https://github.com/hpcc-systems/HPCC-Platform/pull/15155/files/bda698a7bd496408a28db7bc31abc0c66d9786ff#r667769667
..is still applicable, there can only be one key value in the dict, the tmergeOverwrite will ensure that it is last.

HPCCSmoketest commented:
Process of PR-15155, label: HPCC-26155-allow-placements-appending starts now.
The reason of this test is: Forced to re-test.
Commit ID: 6df06c774f33386a05051c8bfde5f70a912e514b
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-241.ca-central-1.compute.internal
 Git:  git version 2.9.5
Build only for containerized environment.

jakesmith reviewed (approved):
@xwang2713 - looks fine, please squash

xwang2713 commented:
Summary of the changes:
1) apply {{ toYaml .me }}
2) replace merge with mergeOverwrite
3) update placements.md for nodeSelector

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-241.ca-central-1.compute.internal
 Git:  git version 2.9.5
Sha: 6df06c774f33386a05051c8bfde5f70a912e514b
Containerized:True
Build: success
Milestone:Install hpccsystems-platform-community_8.2.7-closedown0.el7.x86_64.rpm
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  494 sec (00:08:14) |  111 sec (00:01:51) |   20 sec (00:00:20) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  633 sec (00:10:33) |

jakesmith reviewed (commented):
@xwang2713 - please squash

xwang2713 commented:
Squashed.

HPCCSmoketest commented:
Process of PR-15155, label: HPCC-26155-allow-placements-appending starts now.
The reason of this test is: Forced to re-test.
Commit ID: 0b3986b506ffe47ddd6dae79cd705f408d3bc6c4
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-243.ca-central-1.compute.internal
 Git:  git version 2.9.5
Build only for containerized environment.

jakesmith reviewed (approved):
@xwang2713 - looks fine afaics

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-243.ca-central-1.compute.internal
 Git:  git version 2.9.5
Sha: 0b3986b506ffe47ddd6dae79cd705f408d3bc6c4
Containerized:True
Build: success
Milestone:Install hpccsystems-platform-community_8.2.7-closedown0.el7.x86_64.rpm
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  496 sec (00:08:16) |  104 sec (00:01:44) |   20 sec (00:00:20) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  628 sec (00:10:28) |

g-pan commented on helm/hpcc/templates/_helpers.tpl:
```diff
@@ -1061,11 +1061,30 @@ Pass in dict with root, category.  optional name to restrict it to a single name
 {{/*
 Create placement related settings
-Pass in dict with placement
+Pass in dict with me for current placements and dict with new for the new placements
```
This is very minor - however this sentence confused me a little - since I don't really understand all these variables, but if this is appropriate language I would rather say something like pass in dict with the me prefix (if that is an appropriate term) for current placements and dict with the new prefix for the new placements

g-pan reviewed (commented):
One very minor comment on line 1064

xwang2713 commented:
Fixed typo and rebased.

HPCCSmoketest commented:
Process of PR-15155, label: HPCC-26155-allow-placements-appending starts now.
The reason of this test is: Forced to re-test.
Commit ID: 601697236eed3779efd0bc35b0b338fae630f85a
Estimated completion time is 0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-6.ca-central-1.compute.internal
 Git:  git version 2.9.5
Build only for containerized environment.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-6.ca-central-1.compute.internal
 Git:  git version 2.9.5
Sha: 601697236eed3779efd0bc35b0b338fae630f85a
Containerized:True
Build: success
Milestone:Install hpccsystems-platform-community_8.2.7-closedown0.el7.x86_64.rpm
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  495 sec (00:08:15) |  107 sec (00:01:47) |   19 sec (00:00:19) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  629 sec (00:10:29) |
